pkg/controllers/routereflector/bgppeer.go

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2020 IBM Corporation All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routereflector
+
+import (
+	"context"
+
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	calicoClient "github.com/projectcalico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/libcalico-go/lib/options"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type bgpPeer interface {
```

Comment:
The main reason of this extra layer is the code follows dependency injection design pattern (this is my fetish). This makes code really easy to test. In this way the method like this https://github.com/projectcalico/kube-controllers/pull/514/files#diff-fc4d3fb446ff22ad078c51b10a48a4093a7b7a5967c24acf4640a7d629fcda46R140 can be tested like a flick because logging it's only side effect. It isn't coupled to any implementation.

Reply:
If we use `interface{}` in "objectSaver" at one point here: we have to cast it to `*apiv3.BGPPeer` which i wanted to avoid. With interface we loose type information. But this is my preference. What do you think?